### ISSUE TITLE
custom gateway not working: istio-autogenerated-k8s-ingress seems hardcoded

### ISSUE THREAD

**Issue by Morriz** [istio/istio#13113]
**Describe the bug**
It seems custom gateways do not work when going through the [ingress-certmgr tutorial](https://istio.io/docs/examples/advanced-gateways/ingress-certmgr/).

Only the gateway that I start with is honored, and it has the name `istio-autogenerated-k8s-ingress` (from helm install). If I copy that gw, and only modify that name, and replace the gateway for the one with the name mod, no routes are setup.

**Expected behavior**

To be able to deploy custom gw and have changes automatically picked up

**Steps to reproduce the bug**

`kubectl -n istio-system edit gw istio-autogenerated-k8s-ingress`
 
copy the manifest to another file, and modify the resource name, copy in memory

delete the only gw:

`kubectl -n istio-system delete gw  istio-autogenerated-k8s-ingress`

now create the modified one:

`pbpaste|kubectl -n istio-system apply -f-`

watch in other window:

`kubectl -n istio-system logs -f deployment.extensions/istio-ingressgateway -c istio-proxy --tail=10`

try to reach the service again (which now 404s): 
`curl http://$INGRESS_DOMAIN/hello`

**Version**
istio: 1.1.1
kube: v1.10.11

**Installation**
see tutorial

**Environment**
osx: docker for mac kubernetes

**Comment by robinbryce** [comment]
The istio-autogenerated-k8s-ingress gateway is special. Its name is significant to the implementation of Pilot and Galley (At least up to Istio 1.1.0) 

TL;DR By deleting this gateway and then 'restoring' it with a different name, you are actually disabling istio's support for traditional kubernetes Ingress resources. 

istio-autogenerated-k8s-ingress is created as a consequence of --set global.k8sIngress.enabled=true. The helm chart for istio has a conditional section defining this gateway and which is enabled by that variable[1]

Pilot's implementation[2] contains the following
```
// IstioIngressGatewayName is the internal gateway name assigned to ingress
IstioIngressGatewayName = "istio-autogenerated-k8s-ingress"
```
Both Galley and Pilot have numerous references implying some kind of mapping from what I presume would be k8s 'Ingress' resources to Istio's VirtualServices.

Version:
Istio: 1.1.0
Kube: 1.10.11
Environment:
Azure AKS

1. https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/charts/gateways/templates/preconfigured.yaml

**Comment by Morriz** [comment]
thanks a lot! that explains it...I lost many hours over this. I tried figuring out why sometimes it favored my old ingress resources over a virtualservice. 

I think this should be covered very well in the docs!

**Comment by Morriz** [comment]
But that still does not explain why my virtualservices don't get picked up....I tried exposing the telemetry services according to the manual, but they did not get picked up

**Comment by Morriz** [comment]
no other gateway other than the `istio-autogenerated-k8s-ingress` has worked so far

**Comment by jmagnusson** [comment]
Lost a lot of hours on this as well... Any chance we could get this covered in the docs?

**Comment by sudomann** [comment]
Chiming in to say this should be better reflected in docs as well. Thankfully I found this before wasting hours to figure out what was going on. In my case: after I would create my custom gateway, it was detected/used if I did one of the following:

- Added `istio-autogenerated-k8s-ingress` to the `spec.gateways` list of my `VirtualService`.
- Deleted the `istio-autogenerated-k8s-ingress` gateway
  - And then I found that recreating it (I had saved it to a `yaml` file) after deleting didn't cause it to interfere with my custom gateway anymore

**Comment by gustavovalverde** [comment]
Related: https://github.com/istio/istio/issues/15277

**Comment by gustavovalverde** [comment]
I'm kind of lost here.

I've been dealing with **cert-manager** to make it work with **http01**. The only way to make it work was adding the _istio-autogenerated-k8s-ingress_ allowing all traffic to port 80:

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: istio-autogenerated-k8s-ingress
  namespace: istio-system
  labels:
    app: ingressgateway
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      protocol: HTTP2
      name: http
    hosts:
    - "*"
---
```

This, with a custom DestinationRule to disable mTLS for "cert-manager" _while validating_ the certificate, allowed me to make less steps to get a certificate from a different domain.
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: cert-htt01-challenge
  namespace: istio-system
spec:
  host: "*.istio-system.svc.cluster.local"
  trafficPolicy:
    tls:
      # keeping the same mTLS mode as  `default` DestinationRule in istio-system
      mode: DISABLE
    portLevelSettings:
    - port:
        # CertManager generate services to perform the challenge on port 8089
        # it looks so far no other services in istio-system use this port
        number: 8089
      tls:
        mode: DISABLE
```

But now I'm getting **random** redirections because I have 2 gateways, the _istio-autogenerated-k8s-ingress_ and my custom gateway which has `httpsRedirect: true` and routes to the real services.

The only way to fix this is deleting the _istio-autogenerated-k8s-ingress_ gateway, but if need another certificate I'll have to re-create it again, then delete it, and so on for each new certificate.

Is there a way around this?

**Comment by perrymana** [comment]
I gave up trying to get **http** verification working with Istio, ended up going **dns** validation instead which I haven't had any problems with. This does away with the need for this custom gateway.

**Comment by spacejam101** [comment]
Any update on this? Also currently having the same issue
